# SICP Chapter 2

In [1]:
;; from book -> not technically correct as will return negative
(define (gcd a b)
  (if (= b 0)
      a
      (gcd b (remainder a b))))

;; construct rational number reduced to lowest terms
(define (make-rat n d) 
  (let ((g (gcd n d)))
    (cons (/ n g)
          (/ d g))))

(define (numer x) (car x))
(define (denom x) (cdr x))

(define (add-rat x y)
  (make-rat (+ (* (numer x) (denom y))
               (* (numer y) (denom x)))
            (* (denom x) (denom y))))

(define (sub-rat x y)
  (make-rat (- (* (numer x) (denom y))
               (* (numer y) (denom x)))
            (* (denom x) (denom y))))

(define (mul-rat x y)
  (make-rat (* (numer x) (numer y))
            (* (denom x) (denom y))))

(define (div-rat x y)
  (make-rat (* (numer x) (denom y))
            (* (denom x) (numer y))))

(define (equal-rat? x y)
  (= (* (numer x) (denom y))
     (* (numer y) (denom x))))

(define (print-rat x)
  (newline)
  (display (numer x))
  (display "/")
  (display (denom x)))

In [2]:
(define half (make-rat 1 2))
(print-rat half)


1/2

In [3]:
(define fifth (make-rat 1 5))
(print-rat fifth)


1/5

In [4]:
(define third (make-rat 1 3))
(print-rat third)


1/3

In [5]:
(print-rat (mul-rat half fifth))


1/10

In [6]:
(print-rat (add-rat half fifth))


7/10

In [7]:
(print-rat (add-rat half half))


1/1

In [8]:
(print-rat (add-rat third third))


2/3

## 2.1

In [9]:
(define (make-rat n d) 
  (let ((g (abs (gcd n d)))) ; abs to correct for negative return from gcd
    (if (< d 0)
        (cons (/ (* n -1) g) 
              (/ (* d -1) g))
        (cons (/ n g) 
              (/ d g)))))
    

In [10]:
(print-rat (make-rat 1 2))


1/2

In [11]:
(print-rat (make-rat -1 2))


-1/2

In [12]:
(print-rat (make-rat 1 -2))


-1/2

## 2.2

In [13]:
(define (make-point x y)
  (cons x y))
(define (x-point point)
  (car point))
(define (y-point point)
  (cdr point))

(define (make-segment start-point end-point)
  (cons start-point end-point))
(define (start-segment segment)
  (car segment))
(define (end-segment segment)
  (cdr segment))

(define (midpoint-segment segment)
  (define (avg x y) (/ (+ x y) 2.0))
  (let ((start (start-segment segment)) (end (end-segment segment)))
    (make-point
     (avg (x-point start) (x-point end))
     (avg (y-point start) (y-point end)))))

(define (print-point p)
  (newline)
  (display "(")
  (display (x-point p))
  (display ",")
  (display (y-point p))
  (display ")"))

In [14]:
(print-point (make-point 1 2))


(1,2)

In [15]:
(define start (make-point 1 2))
(define end (make-point 2 4))

(define seg (make-segment start end))
(print-point (midpoint-segment seg))


(1.5,3.0)

## 2.3

In [16]:
;; opposite corners representation
(define (make-rect bottom-left top-right)
  (cons bottom-left top-right))
(define (rect-width rect)
  (abs (- (x-point (car rect)) (x-point (cdr rect)))))
(define (rect-height rect)
  (abs (- (y-point (car rect)) (y-point (cdr rect)))))

(define (rect-perim rect)
  (* 2 (+ (rect-width rect) (rect-height rect))))

(define (rect-area rect)
  (* (rect-width rect) (rect-height rect)))

In [17]:
(define rect (make-rect (make-point 1 2) (make-point 3 5)))

(rect-width rect)

2

In [18]:
(rect-height rect)

3

In [19]:
(rect-perim rect)

10

In [20]:
(rect-area rect)

6

In [21]:
;; bottom left corner + width/height representation
(define (make-rect bottom-left width height)
  (cons bottom-left (cons width height)))
(define (rect-width rect)
  (car (cdr rect)))
(define (rect-height rect)
  (cdr (cdr rect)))

In [22]:
(define rect (make-rect (make-point 1 2) 2 3))

In [23]:
(rect-perim rect)

10

In [66]:
(rect-area rect)

6

## 2.4

In [32]:
;; returns function that takes 2 arguments and applies it to x y
(define (cons x y)
  (lambda (m) (m x y)))

;; z takes a 2 argument function that is passed p q in order
;; returning first arg p implements car function
(define (car z)
  (z (lambda (p q) p)))

;; z takes a 2 argument function that is passed p q in order
;; returning sencond arg q produces cdr
(define (cdr z)
  (z (lambda (p q) q)))

In [27]:
(car (cons 1 2))

1

In [33]:
(cdr (cons 1 2))

2

Applicative-order evaluation to verify `(car (cons x y))` yields `x` for any objects `x` and `y`:
```scheme
(car (cons (x y)))
(car (lambda (m) (m x y)))
((lambda (m) (m x y)) (lambda (p q) p))
((lambda (p q) p) x y)
x
```

## 2.5

Represent pair of nonnegative integers $a, b$ as the integer that is the product $2^a3^b$

2 and 3 are prime -> $2^a3^b$ will give unique values for each all combinations of $a, b$ due to the [Fundamental Theorem of Arithmetic](http://mathworld.wolfram.com/FundamentalTheoremofArithmetic.html)

$a$ can be calculated from the product by the number of times it is evenly divisible by 2.

$b$ can be calculated from the product by the the number of times it is evenly divisible by 3.

Example:

$a=2\ ,b=3$

$2^a3^b=2^2\cdot3^3=108$

Calculating $a$:

$\frac{108}{2}=54$

$\frac{54}{2}=27$

$\frac{27}{2}=13.5$

2 even divisions by 2:

$a=2$

Calculating $b$:

$\frac{108}{3}=36$

$\frac{36}{3}=12$

$\frac{12}{3}=4$

$\frac{4}{3} = 1.3333333$

3 even divisions by 3:

$b=3$




In [11]:
(define (cons a b)
  (* (expt 2 a)
     (expt 3 b)))

; number of times n is evenly divisible by d
(define (num-div n d)
  (define (iter x count)
    (if (= 0 (remainder x d))
        (iter (/ x d) (+ 1 count))
        count))
  (iter n 0))

(define (car x)
  (num-div x 2))

(define (cdr x)
  (num-div x 3))

In [12]:
(car (cons 2 3))

2

In [13]:
(cdr (cons 2 3))

3

## 2.6

In [17]:
(define zero (lambda (f) (lambda (x) x)))

(define (add-1 n)
  (lambda (f) (lambda (x) (f ((n f) x)))))

Substitution to find definition of `one`:
```scheme
(add-1 zero)
(lambda (f) (lambda (x) (f ((zero f) x))))
(lambda (f) (lambda (x) (f ((lambda (x) x) x))))
(lambda (f) (lambda (x) (f x)))
```

In [19]:
;; applies given procedure once
(define one (lambda (f) (lambda (x) (f x))))

Substitution to find definition of `two`:
```scheme
(add-1 one)
(lambda (f) (lambda (x) (f ((one f) x))))
(lambda (f) (lambda (x) (f ((lambda (x) (f x)) x))))
(lambda (f) (lambda (x) (f (f x))))
```

In [20]:
;; applies given procedure twice
(define two (lambda (f) (lambda (x) (f (f x)))))

In [24]:
;; 'wrap' n in m function calls
(define (church+ m n)
  (lambda (f) (lambda (x) ((m f) ((n f) x)))))

In [38]:
;; easy to see number of function applications with increment procedure
(define (increment n)
  (+ n 1))

In [39]:
((zero increment) 0)

0

In [40]:
((zero increment) 1)

1

In [41]:
((zero increment) 2)

2

In [42]:
((one increment) 0)

1

In [43]:
((one increment) 1)

2

In [44]:
((one increment) 2)

3

In [45]:
((two increment) 0)

2

In [46]:
((two increment) 1)

3

In [47]:
((two increment) 2)

4

In [ ]:
(define three (church+ one two))

((three increment) 0)

## 2.7

In [1]:
;; procedures given in question description
(define (make-interval a b) (cons a b))

(define (add-interval x y)
  (make-interval (+ (lower-bound x)
                    (lower-bound y))
                 (+ (upper-bound x)
                    (upper-bound y))))

;; find minimum and maximum of products of the bounds
(define (mul-interval x y)
  (let ((p1 (* (lower-bound x)
               (lower-bound y)))
        (p2 (* (lower-bound x)
               (upper-bound y)))
        (p3 (* (upper-bound x)
               (lower-bound y)))
        (p4 (* (upper-bound x)
               (upper-bound y))))
    (make-interval (min p1 p2 p3 p4)
                   (max p1 p2 p3 p4))))
(define (div-interval x y)
  (mul-interval x
                (make-interval
                 (/ 1.0 (upper-bound y))
                 (/ 1.0 (lower-bound y)))))

In [2]:
(define (lower-bound x) (min (car x) (cdr x)))
(define (upper-bound x) (max (car x) (cdr x)))

In [3]:
;; test with a=lower, b=upper
(define x (make-interval 10 20))

(lower-bound x)

10

In [4]:
(upper-bound x)

20

In [5]:
;; test with a=upper, b=lower
(define y (make-interval 20 10))

(lower-bound y)

10

In [6]:
(upper-bound y)

20

## 2.8
Difference between intervals `x, y`:
- Minimum = lower bound `x` subtract upper bound `y` 
- Maximum = upper bound `x` subtract lower bound `y`

In [7]:
(define (sub-interval x y)
  (make-interval 
   (- (lower-bound x) (upper-bound y))
   (- (upper-bound x) (lower-bound y))))

(define x (make-interval 10 20))
(define y (make-interval 1 5))
(sub-interval x y)

(5 . 19)

## 2.9

`add-interval`:

$[a,b]+[c,d]=[(a+c), (b+d)]$

`width`:

$width([a,b])=\frac{b-a}{2}$

`width` of `add-interval`:

$width([a,b]+[c,d])=width([(a+c), (b+d)])$

$=\frac{(b+d)-(a+c)}{2}$

sum of `width` of two intervals:

$width([a,b]) + width([c,d])=\frac{b-a}{2}+\frac{d-c}{2}$

$=\frac{(b-a)+(d-c)}{2}$

$=\frac{(b+d)-(a+c)}{2}$

`width` of `add-interval` is equal to the sum of `with` of the same two intervals. Therefore the width of the the sum of two intervals is a function only of the widthds of the intervals being added.

In [8]:
(define (width x)
  (/ (- (upper-bound x) (lower-bound x)) 2))

(define a (make-interval 10 20))
(define b (make-interval 30 40))
(define c (make-interval 1 3))

In [9]:
(width a)

5

In [10]:
(width b)

5

In [11]:
(width c)

1

In [12]:
(width (add-interval a b))

10

Intervals `a` and `b` have the same width. Multiplying each of them by `c` gives produce intervals with different widths. Therefore the product of two intervals has a width that is not a function solely of the widths of said intervals.

In [13]:
(width (mul-interval a c))

25

In [14]:
(width (mul-interval b c))

45

## 2.10

In [15]:
(define (zero-span? x)
  (and (<= (lower-bound x) 0)
       (>= (upper-bound x) 0)))

;; throws an error if divide by 0
(define (div-interval x y)
  (if (zero-span? y)
      (error "div-interval" "Cannot divide by interval with span 0")
      (mul-interval x
                (make-interval
                 (/ 1.0 (upper-bound y))
                 (/ 1.0 (lower-bound y))))))

(div-interval (make-interval 10 20) (make-interval -1 1))


Traceback (most recent call last):
  File "In [15]", line 14, col 1, in 'div-interval'
  File "In [15]", line 8, col 7, in 'error'
  File "In [15]", line 8, col 7
RunTimeError: Error in 'div-interval': Cannot divide by interval with span 0



## 2.11
`(mul-interval x y)` produces an interval where the lower bound is the minimum product of `x` and `y`'s bounds  and the upper bound is the maximum product of `x` and `y`'s bounds.

Upper and lower bounds of an interval can be both positive, both negative or negative lower with positive upper i.e. `[+,+]`, `[-,-]` or `[-,+]`
- Can't have positive lower and negative upper as the lower bound would be less than the upper bound

For `mul-interval` this gives the combinations:
```
[+,+] * [+,+]
[+,+] * [-,-]
[+,+] * [-,+]

[-,-] * [-,-]
[-,-] * [+,+]
[-,-] * [-,+]

[-,+] * [-,+]
[-,+] * [+,+]
[-,+] * [-,-]
```

More than 2 multiplications are required for the case `[-,+] * [-,+]` as the product of the two negative lower bounds could be larger than that of the upper bounds.

In all other cases only two multiplications are required as there is only one combination of products that can produce the correct result
- e.g. `[+,+] * [+,+]` -> multiply two upper bounds and multiply two lower bounds.

In [16]:
(define (pos? x) (>= x 0))
(define (neg? x ) (<= x 0))

(define (mul-interval x y)
  (let ((lx (lower-bound x))
        (ux (upper-bound x))
        (ly (lower-bound y))
        (uy (upper-bound y)))
    (cond ((and (pos? lx)
                (pos? ux)
                (pos? ly)
                (pos? uy))
           ;; [+, +] * [+, +]
           (make-interval (* lx ly) (* ux uy)))
          ((and (pos? lx)
                (pos? ux)
                (neg? ly)
                (pos? uy))
           ;; [+, +] * [-, +]
           (make-interval (* ux lx) (* ux uy)))
          ((and (pos? lx)
                (pos? ux)
                (neg? ly)
                (neg? uy))
           ;; [+, +] * [-, -]
           (make-interval (* ux lx) (* lx uy)))
          ((and (neg? lx)
                (pos? ux)
                (pos? ly)
                (pos? uy))
           ;; [-, +] * [+, +]
           (make-interval (* lx uy) (* ux uy)))
          ((and (neg? lx)
                (pos? ux)
                (neg? ly)
                (pos? uy))
           ;; [-, +] * [-, +]
           (make-interval (min (* ux ly) (* lx uy))
                          (max (* lx ly) (* ux uy))))
          ((and (neg? lx)
                (pos? ux)
                (neg? ly)
                (neg? uy))
           ;; [-, +] * [-, -]
           (make-interval (* ux ly) (* lx ly)))
          ((and (neg? lx)
                (neg? ux)
                (pos? ly)
                (pos? uy))
           ;; [-, -] * [+, +]
           (make-interval (* lx ux) (* ux ly)))
          ((and (neg? lx)
                (neg? ux)
                (neg? ly)
                (pos? uy))
           ;; [-, -] * [-, +]
           (make-interval (* lx ux) (* lx ly)))
          ((and (neg? lx)
                (neg? ux)
                (neg? ly)
                (neg? uy))
           ;; [-, -] * [-, -]
           (make-interval (* ux uy) (* lx ly))))))

In [17]:
(define a (make-interval 1 2))
(define b (make-interval -1 2))
(define c (make-interval -1 -2))

In [18]:
(mul-interval a a)

(1 . 4)

In [19]:
(mul-interval a b)

(2 . 4)

In [20]:
(mul-interval a c)

(2 . -1)

In [21]:
(mul-interval b a)

(-2 . 4)

In [22]:
(mul-interval b b)

(-2 . 4)

In [23]:
(mul-interval b c)

(-4 . 2)

In [24]:
(mul-interval c a)

(2 . -1)

In [25]:
(mul-interval c b)

(2 . 2)

In [65]:
(mul-interval c c)

(1 . 4)

## 2.12

In [27]:
;; definitions given in question
(define (make-center-width c w)
  (make-interval (- c w) (+ c w)))

(define (center i)
  (/ (+ (lower-bound i)
        (upper-bound i))
     2))

(define (width i)
  (/ (- (upper-bound i)
        (lower-bound i))
     2))

In [62]:
;; using make-interval directly
(define (make-center-percent c tolerance)
  (let ((unc (* (/ c 100) tolerance)))
    (make-interval (- c unc) (+ c unc))))

;; using provided make-center-width
(define (make-center-percent c tolerance)
  (make-center-width c (* (/ c 100) tolerance)))

(define (percent i)
  (* (/ (width i) (center i)) 100.0))

In [63]:
(define i (make-center-percent 10 25))
i

(15/2 . 25/2)

In [64]:
(center i)

10

In [65]:
(percent i)

25.0

## 2.13

Interval multiplication:

$[a,b]\cdot[c,d] = [\min(ac,ad,bc,bd), \max(ac,ad,bc,bd)]$

Assuming all values are positive:

$[a,b]\cdot[c,d] = [ac, bd]$

Interval represented by its center $c_i$ and tolerance percentage $p_i$ :

$i = [c_i-\frac{c_i\cdot p_i}{100},c_i+\frac{c_i\cdot p_i}{100}]$

$= [c_i\frac{1-p_i}{100},c_i\frac{1+p_i}{100}]$

Multiplying two intervals $i,j$:

$ij=[c_i\frac{1-p_i}{100},c_i\frac{1+p_i}{100}]\cdot[c_j\frac{1-p_j}{100},c_j\frac{1+p_j}{100}]$

$=[c_i\frac{1-p_i}{100}\cdot c_j\frac{1-p_j}{100}, c_i\frac{1+p_i}{100}\cdot c_j\frac{1+p_j}{100}]$

$=[c_ic_j(1-\frac{p_i+p_j}{100}+\frac{p_ip_j}{10000}), c_ic_j(1+\frac{p_i+p_j}{100}+\frac{p_ip_j}{10000})]$

Percentage tolerances are small ->$\frac{p_ip_j}{10000}$ terms will be small -> approximate as:

$ij=[c_ic_j(1-\frac{p_i+p_j}{100}), c_ic_j(1+\frac{p_i+p_j}{100})]$

Percentage tolerance of the product = $p_i+p_j$
- **sum of the tolerances of the two intervals**

In [67]:
;; test with small tolerances
(define i (make-center-percent 100 2))
(define j (make-center-percent 80 1))

(define ij (mul-interval i j))
;; approximately 2+1=3
(percent ij)

2.9994001199760048

## 2.14

In [68]:
;; procedures from question
(define (par1 r1 r2)
  (div-interval
   (mul-interval r1 r2)
   (add-interval r1 r2)))

(define (par2 r1 r2)
  (let ((one (make-interval 1 1)))
    (div-interval
     one
     (add-interval 
      (div-interval one r1)
      (div-interval one r2)))))

In [85]:
;; demonstrate different results between par1 and par2
(define A (make-center-percent 100 3))
(define B (make-center-percent 200 2))

(par1 A B)

(61.928338762214985 . 71.71331058020478)

In [86]:
(par2 A B)

(64.88737201365187 . 68.44299674267101)

In [87]:
;; investigate behaviour with arithmetic expressions
(define A/A (div-interval A A))
(define A/B (div-interval A B))

(center A/A)

1.0018016214593133

In [88]:
(percent A/A)

5.994604855629938

In [89]:
(center A/B)

0.500500200080032

In [90]:
(percent A/B)

4.997001798920647

The expected center of dividing an interval by itself is 1, however in the investigation above, `(center A/A)` returns `1.0018016214593133` - an approximation to 1.

## 2.15
The question states that `par1` and `par2` use two different but algebraically equivalent formulae:
- `par1` uses the formula $\frac{R_1R_2}{R_1+R_2}$
- `par2` uses the formula $\frac{1}{\frac{1}{R_1}+\frac{1}{R_2}}$

Proof of algrebraic equivalence:

Multiply by $\frac{R_1}{R_1}$ and $\frac{R_2}{R_2}$. These are both equivalent to 1 and so the formula is still valid:

$\frac{1}{\frac{1}{R_1}+\frac{1}{R_2}}=\frac{R_1}{R_1}\frac{R_2}{R_2}\frac{R_1R_2}{R_1+R_2}$

$=\frac{R_1R_2}{R_1R_2(\frac{1}{R_1}+\frac{1}{R_2})}$

$=\frac{R_1R_2}{\frac{R_1R_2}{R_1}+\frac{R_1R_2}{R_2}}$

$=\frac{R_1R_2}{R_1+R_2}$

The transformation assumes that $\frac{R_1}{R_1}$ and $\frac{R_2}{R_2}$ equal 1. However it was shown in question 2.14 that dividing an interval by itself will produce a result with a center that is an **approximation** to 1 - introducing the descrepancy between `par1` and `par2`. Therefore, tighter error bounds will be achieved by avoiding repetition of variables that represent uncertain numbers - **Eva Lu Ator is correct**.

## 2.17

In [1]:
(define (last-pair l)
  (if (null? (cdr l))
      l
      (last-pair (cdr l))))

In [2]:
(last-pair (list 23 72 149 34))

(34)

## 2.18

In [12]:
;; given in question
(define (append list1 list2)
  (if (null? list1)
      list2
      (cons (car list1)
            (append (cdr list1)
                    list2))))

;; reverse by appending car of a list to the reverse of cdr of the list
(define (reverse l)
  (if (null? l)
      l
      (append (reverse (cdr l)) (list (car l)))))

(reverse (list 1 4 9 16 25))

(25 16 9 4 1)

## 2.19

In [3]:
;; given in question
(define us-coins 
  (list 50 25 10 5 1))

(define (cc amount coin-values)
  (cond ((= amount 0)
         1)
        ((or (< amount 0)
             (no-more? coin-values))
         0)
        (else
         (+ (cc
             amount
             (except-first-denomination
              coin-values))
            (cc
             (- amount
                (first-denomination
                 coin-values))
             coin-values)))))

(define (first-denomination coin-values)
  (car coin-values))

(define (except-first-denomination coin-values)
  (cdr coin-values))

(define (no-more? coin-values)
  (null? coin-values))

In [27]:
(cc 100 us-coins)

292

##### The order of the list `coin-values` will **not effect** the answer produced by `cc`
- Each sub-list of `coin-values` is evaluated recursively after subtracting the value of the first coin from the amount
- All possible combinations are computed regardless of the order of `coin-values`

In [4]:
(cc 100 (reverse us-coins))

292

In [6]:
(define us-coins-shuffled (list 25 5 50 1 10))
(cc 100 us-coins-shuffled)

292

## 2.20

In [21]:
;; check parity of first element x
;; filter list using function that determines if input is that parity
(define (same-parity x . y)
  (define (iter z filter)
    (cond ((null? z) z)
          ((filter (car z))
            ;; parity matches append first item, check next
           (cons (car z) (iter (cdr z) filter)))
          ;; parity mismatch, check next
          (else (iter (cdr z) filter))))
  (iter (cons x y) (if (even? x) even? odd?)))

In [22]:
(same-parity 1 2 3 4 5 6 7)

(1 3 5 7)

In [23]:
(same-parity 2 3 4 5 6 7)

(2 4 6)

## 2.21

In [22]:
(define (map proc items)
  (if (null? items)
      items
      (cons (proc (car items))
            (map proc (cdr items)))))

(define (square-list items)
  (if (null? items)
      items
      (cons (* (car items) (car items)) 
            (square-list (cdr items)))))

(square-list (list 1 2 3 4))

(1 4 9 16)

In [23]:
(define (square-list items)
  (map (lambda (x) (* x x)) items))

In [24]:
(square-list (list 1 2 3 4))

(1 4 9 16)

## 2.22

In [38]:
(define (square x) (* x x))

;; Result list is reversed
(define (square-list items)
  (define (iter things answer)
    (if (null? things)
        answer
        (iter (cdr things)
              ;; square item from front of input list and append to 
              ;; front of answer list
              (cons (square (car things))
                    answer))))
  ;; start with empty list due to implementation details of nil value
  (iter items (list)))

(square-list (list 1 2 3 4))

(16 9 4 1)

On each iteration, an item is taken from the front of the original list, is squared and then appends the result to the *front* of the answer list. The last item to be squared from the original list will therfore end up as the first item in the answer list 

In [39]:
(define (square-list items)
  (define (iter things answer)
    (if (null? things)
        answer
        (iter (cdr things)
              ;; pass list as first arg to cons
              (cons answer 
                    (square
                     (car things))))))
  (iter items (list)))

(square-list (list 1 2 3 4))

((((() . 1) . 4) . 9) . 16)

Flipping the order of arguments to `cons` creates a pair containing the answer list and the next squared integer. It does *not* append the integer onto the list. Each iteration combines the list from the previous step with the next integer result and wraps it in a new list.

## 2.23

In [44]:
(define (for-each proc items)
  (cond ((null? items)
         #t)
        (else (proc (car items))
              (for-each proc (cdr items)))))

(for-each 
 (lambda (x) (newline) (display x))
 (list 1 2 3 4))


1
2
3
4

#t

## 2.24 TODO - ADD DIAGRAMS

In [45]:
(list 1 (list 2 (list 3 4)))

(1 (2 (3 4)))

## 2.25

In [47]:
(define a (list 1 3 (list 5 7) 9))
a

(1 3 (5 7) 9)

In [57]:
(car (cdr (car (cdr (cdr a)))))

7

In [48]:
(define b (list (list 7)))
b

((7))

In [56]:
(car (car b))

7

In [49]:
(define c (list 1 (list 2 (list 3 (list 4 (list 5 (list 6 7)))))))
c

(1 (2 (3 (4 (5 (6 7))))))

In [72]:
(cdr (car (cdr (car (cdr (car (cdr (car (cdr (car (cdr c)))))))))))

(7)

## 2.26

In [73]:
(define x (list 1 2 3))
(define y (list 4 5 6))

In [76]:
(append x y)

(1 2 3 4 5 6)

In [77]:
(cons x y)

((1 2 3) 4 5 6)

In [78]:
(list x y)

((1 2 3) (4 5 6))

## 2.27

In [83]:
;; from 2.18
(define (reverse l)
  (if (null? l)
      l
      (append (reverse (cdr l)) (list (car l)))))

(define (deep-reverse l)
  (cond ((null? l) (list))
        ((pair? (car l))
         (append (deep-reverse (cdr l))
                 (list (deep-reverse (car l)))))
        (else (append (deep-reverse (cdr l))
                      (list (car l))))))

In [84]:
(define x (list (list 1 2) (list 3 4)))
x

((1 2) (3 4))

In [85]:
(reverse x)

((3 4) (1 2))

In [88]:
(deep-reverse x)

((4 3) (2 1))

## 2.28

In [103]:
;; traverse tree->append each node to list result,
(define (fringe t)
  (cond ((null? t) (list))
        ((not (pair? t)) (list t))
        (else (append (fringe (car t))
                      (fringe (cdr t))))))

In [104]:
(define x (list (list 1 2) (list 3 4)))
x

((1 2) (3 4))

In [105]:
(fringe x)

(1 2 3 4)

## 2.29

In [111]:
(define (make-mobile left right)
  (list left right))

(define (make-branch len structure)
  (list len structure))

(define (left-branch mobile)
  (car mobile))

(define (right-branch mobile)
  (car (cdr mobile)))

(define (branch-length branch)
  (car branch))

(define (branch-structure branch)
  (car (cdr branch)))

;; mobile total weight = sum of left + right branch weight
;; branch weight = value of weight if not a mobile, or recursive if branch
;; holds a mobile
(define (branch-weight branch)
  (if (pair? (branch-structure branch))
      (total-weight (branch-structure branch))
      (branch-structure branch)))

(define (total-weight mobile)
  (+ (branch-weight (left-branch mobile))
     (branch-weight (right-branch mobile))))

(define (branch-torque branch)
  (* (branch-length branch)
     (branch-weight branch)))

(define (branch-balanced? branch)
  (if (pair? (branch-structure branch))
      (mobile-balanced? (branch-structure branch))
      #t))
(define (mobile-balanced?)
  (and (= (branch-torque (left-branch mobile))
          (branch-torque (right-branch mobile)))
       (branch-balanced? (left-branch mobile))
       (branch-balanced? (right-branch mobile))))

Changing the representation to:
```scheme
(define (make-mobile left right)
  (cons left right))

(define (make-branch length structure)
  (cons length structure))
```

Would only affect the `right-branch` and `branch-structure` accessors:
```scheme
(define (right-branch mobile)
  (cdr mobile))

(define (branch-structure branch)
  (cdr branch))
```

## 2.30

In [5]:
(define t (list 1
                (list 2 (list 3 4) 5)
                (list 6 7)))
t

(1 (2 (3 4) 5) (6 7))

In [6]:
(define (square-tree tree)
  (cond ((null? tree) (list))
        ((not (pair? tree))
         (* tree tree))
        (else
         (cons (square-tree (car tree))
               (square-tree (cdr tree))))))

In [7]:
(square-tree t)

(1 (4 (9 16) 25) (36 49))

In [8]:
(define (square-tree tree)
  (map (lambda (sub-tree)
         (if (pair? sub-tree)
             (square-tree sub-tree)
             (* sub-tree sub-tree)))
       tree))

In [9]:
(square-tree t)

(1 (4 (9 16) 25) (36 49))

## 2.31

In [12]:
(define (tree-map proc tree)
  (map (lambda (sub-tree)
         (if (pair? sub-tree)
             (tree-map proc sub-tree)
             (proc sub-tree)))
       tree))

(define (square x) (* x x))

(define (square-tree tree)
  (tree-map square tree))

(square-tree t)

(1 (4 (9 16) 25) (36 49))

## 2.32
Set of subsets of a set = **powerset**

Algorithm from [Wikipedia](https://en.wikipedia.org/wiki/Power_set#Algorithms):
- Powerset of empty set = the set containing the empty set
- Powerset of any other set is all the subsets of the set containing a specific element and all the subsets of the set not containing that element


$S=\{1, 2, 3\}$

$P(S)=\{\{\}\{3\}\{2\}\{2,3\}\{1\}\{1,3\}\{1,2\}\{1,2,3\}\}$

In [36]:
(define (subsets s)
  (if (null? s)
      (list ())
      (let ((rest (subsets (cdr s))))
        (append rest (map (lambda (x) (cons (car s) x)) rest)))))

In [37]:
(subsets (list 1 2 3))

(() (3) (2) (2 3) (1) (1 3) (1 2) (1 2 3))

Line `3` of `subsets` represents the base case:
- $P(\{\})=\{\{\}\}$
`subsets` is then called recursively with `(cdr s)`- is all but the first element of the set. The first element of the set - `(car s)` - is then appended to each subset of `(cdr s)`. The recursion stops when `(cdr s)` on line `4` of `subsets` has no elements and the base case is reached.

## 2.33

In [2]:
;; accumulate definition from book
(define (accumulate op initial sequence)
  (if (null? sequence)
      initial
      (op (car sequence)
          (accumulate op
                      initial
                      (cdr sequence)))))

In [20]:
(accumulate + 0 (list 1 2 3))

6

In [6]:
(accumulate * 1 (list 4 5 6))

120

In [10]:
(define (map p sequence)
  (accumulate (lambda (x y) (cons (p x) y))
              (list) sequence))

(define (square x) (* x x))
(map square (list 1 2 3))

(1 4 9)

In [12]:
(define (append seq1 seq2)
  (accumulate cons seq2 seq1))

(append (list 1 2 3) (list 4 5 6))

(1 2 3 4 5 6)

In [15]:
(define (length sequence)
  (accumulate (lambda (x y) (+ y 1)) 0 sequence))

(length (list 1 2 3))

3

## 2.34

### Horner's rule
Evaluation of the polynomial:

$a_nx^n+a_{n-1}x^{n-1}+...+a_1x+a_0$

At a given value of $x$ can be computed as:

$(...(a_nx+a_{n-1})x+...+a_1)x+a_0$

In [18]:
(define
  (horner-eval x coefficient-sequence)
  (accumulate
   (lambda (this-coeff higher-terms)
     (+ (* x higher-terms) this-coeff))
   0
  coefficient-sequence))

In [19]:
(horner-eval 2 (list 1 3 0 5 0 1))

79

## 2.35

In [4]:
;; from book
(define (enumerate-tree tree)
  (cond ((null? tree) (list))
         ((not (pair? tree)) (list tree))
         (else (append
                (enumerate-tree (car tree))
                (enumerate-tree (cdr tree))))))

(define (count-leaves t)
  (accumulate + 0 (map (lambda (x) 1)
                       (enumerate-tree t))))

In [5]:
(count-leaves (cons (list 1 2) (list 3 4)))

4

## 2.36

In [6]:
(define (accumulate-n op init seqs)
  (if (null? (car seqs))
      (list)
      (cons (accumulate op init (map car seqs))
            (accumulate-n op init (map cdr seqs)))))

In [7]:
(define s (list (list 1 2 3)
                (list 4 5 6) 
                (list 7 8 9) 
                (list 10 11 12)))
s

((1 2 3) (4 5 6) (7 8 9) (10 11 12))

In [51]:
(accumulate-n + 0 s)

(22 26 30)

## 2.37

In [1]:
;; representation of a vector
(list 1 2 3 4)

(1 2 3 4)

In [4]:
;; representation of a matrix
(define m (list (list 1 2 3 4)
                (list 4 5 6 6)
                (list 6 7 8 9)))
m

((1 2 3 4) (4 5 6 6) (6 7 8 9))

In [15]:
;; helper function to print a matrix
(define (print-matrix m)
  (map (lambda (row) (display row)(newline)) m))

(print-matrix m)

(1 2 3 4)
(4 5 6 6)
(6 7 8 9)


(<void> <void> <void>)

In [9]:
;; given in book
(define (dot-product v w)
  (accumulate + 0 (map * v w)))

(dot-product (list 1 2 3 4) (list 4 5 6 7))

60

In [10]:
(define (matrix-*-vector m v)
  (map (lambda (row) (dot-product row v)) m))

(matrix-*-vector m (list 1 2 3 4))

(30 56 80)

In [13]:
;; make columns of input matrix rows of output matrix
(define (transpose mat)
  (accumulate-n cons (list) mat))

(print-matrix (transpose m))

(1 4 6)
(2 5 7)
(3 6 8)
(4 6 9)


(<void> <void> <void> <void>)

In [20]:
(define (matrix-*-matrix m n)
  (let ((cols (transpose n)))
    (map (lambda (row) (matrix-*-vector cols row)) m)))

(define n (list (list 1 2 5 8)
                (list 9 5 2 1)
                (list 8 7 3 4)))

(display "m:")(newline)
(print-matrix m)
(display "n:")(newline)
(print-matrix n)
(display "m*n:")(newline)
(print-matrix (matrix-*-matrix m n))

m:
(1 2 3 4)
(4 5 6 6)
(6 7 8 9)
n:
(1 2 5 8)
(9 5 2 1)
(8 7 3 4)
m*n:
(43 33 18 22)
(97 75 48 61)
(133 103 68 87)


(<void> <void> <void>)

## 2.38
```
3/2
1/6
(1 (2 (3 ())))
(((() 1) 2) 3)
```

In [3]:
;; same as accumulate
(define (fold-right op initial sequence)
   (if (null? sequence)
       initial
       (op (car sequence)
           (fold-right op initial (cdr sequence)))))

(define (fold-left op initial sequence)
  (define (iter result rest)
    (if (null? rest)
        result
        (iter (op result (car rest))
              (cdr rest))))
  (iter initial sequence))

In [29]:
(fold-right / 1 (list 1 2 3))

3/2

In [30]:
(fold-left / 1 (list 1 2 3))

1/6

In [31]:
(fold-right list (list) (list 1 2 3))

(1 (2 (3 ())))

In [32]:
(fold-left list (list) (list 1 2 3))

(((() 1) 2) 3)

`op` should be **commutative** for `fold-left` and `fold-right`to produce the same values for any sequence.

The two procedures differ only in the order in which they process the sequence elements. This means that if `op` is not commutative, different results will be produced.

`*` is commutative, therefore `fold-left` and `fold-right` produce the same values:

In [11]:
(fold-right * 1 (list 1 2 3))

6

In [12]:
(fold-left * 1 (list 1 2 3))

6

`/` is *not* commutative, therefore `fold-left` and `fold-right` prduce different values:

In [13]:
(fold-right / 1 (list 1 2 3))

3/2

In [14]:
(fold-left / 1 (list 1 2 3))

1/6

## 2.39

In [9]:
(define (reverse sequence)
  (fold-right
  (lambda (x y) (append y (list x))) (list) sequence))

In [40]:
(reverse (list 1 2 3))

(3 2 1)

In [69]:
(define (reverse sequence)
  (fold-left
   (lambda (x y) (cons y x)) (list) sequence))

In [70]:
(reverse (list 1 2 3))

(3 2 1)

## 2.40

In [4]:
;; calysto scheme missing filter function
(define (filter predicate seq)
  (accumulate (lambda (e rest)
                (if (predicate e)
                    (cons e rest)
                    rest)) 
              ()
              seq))

(define (enumerate-interval low high)
  (if (> low high)
      ()
      (cons low 
            (enumerate-interval
             (+ low 1)
             high))))

(define (flatmap proc seq)
  (accumulate append () (map proc seq)))

(define (prime-sum? pair)
  (prime? (+ (car pair) (cadr pair))))

(define (make-pair-sum pair)
  (list (car pair)
        (cadr pair)
        (+ (car pair) (cadr pair))))

(define (unique-pairs n)
  (flatmap
   (lambda (i)
     (map (lambda (j) (list i j))
          (enumerate-interval 1 (- i 1))))
   (enumerate-interval 1 n)))

(unique-pairs 6)

((2 1) (3 1) (3 2) (4 1) (4 2) (4 3) (5 1) (5 2) (5 3) (5 4) (6 1) (6 2) (6 3) (6 4) (6 5))

In [41]:
(define (prime-sum-pairs n)
  (map make-pair-sum
       (filter prime-sum? (unique-pairs n))))

## 2.41

In [42]:
;; produce ordered triples of distinct positive integers (i,j,k) up to n
(define (ordered-triples n)
  (flatmap (lambda (i)
             (flatmap (lambda (j)
                        (map (lambda (k) (list i j k))
                             (enumerate-interval 1 (- j 1))))
                      (enumerate-interval 1 (- i 1))))
           (enumerate-interval 1 n)))

(ordered-triples 6)

((3 2 1) (4 2 1) (4 3 1) (4 3 2) (5 2 1) (5 3 1) (5 3 2) (5 4 1) (5 4 2) (5 4 3) (6 2 1) (6 3 1) (6 3 2) (6 4 1) (6 4 2) (6 4 3) (6 5 1) (6 5 2) (6 5 3) (6 5 4))

In [43]:
;; true if sum of triple = s
(define (triple-sum? triple s)
  (= (accumulate + 0 triple) s))

(triple-sum? (list 1 2 3) 6)

#t

In [44]:
(triple-sum? (list 1 2 3) 5)

#f

In [45]:
(define (ordered-triple-sum n s)
  (filter (lambda (triple) (triple-sum? triple s)) (ordered-triples n)))

In [46]:
(ordered-triple-sum 5 10)

((5 3 2) (5 4 1))

In [47]:
(ordered-triple-sum 10 5)

()

In [51]:
(ordered-triple-sum 10 6)

((3 2 1))

In [52]:
(ordered-triple-sum 20 8)

((4 3 1) (5 2 1))

## 2.42

In [4]:
(define (make-position row col)
  (cons row col))
(define (position-row position)
  (car position))
(define (position-col position)
  (cdr position))

In [5]:
(define p (make-position 1 2))
p

(1 . 2)

In [6]:
(position-row p)

1

In [7]:
(position-col p)

2

In [8]:
(define (adjoin-position row col positions)
  (append positions (list (make-position row col))))

(define empty-board (list))

In [9]:
(define board (adjoin-position 1 2 empty-board))
board

((1 . 2))

In [10]:
(adjoin-position 3 4 board)

((1 . 2) (3 . 4))

In [11]:
;; true if queens on same col, row or diagonal
(define (checks? q1 q2)
  (or (= (position-row q1) (position-row q2))
      (= (position-col q1) (position-col q2))
      (= (abs (- (position-row q1) (position-row q2)))
         (abs (- (position-col q1) (position-col q2))))))

In [12]:
(checks? (make-position 1 2) (make-position 1 3))

#t

In [13]:
(checks? (make-position 1 2) (make-position 3 2))

#t

In [14]:
(checks? (make-position 1 2) (make-position 5 6))

#t

In [15]:
(checks? (make-position 1 2) (make-position 6 3))

#f

In [16]:
(define (safe? k positions)
  (let ((kth-q (list-ref positions (- k 1)))
        (rest-q (filter (lambda (q)
                          (not (= k (position-col q))))
                        positions)))
    (define (iter q positions)
      (or (null? positions)
          (and (not (checks? q (car positions)))
               (iter q (cdr positions)))))
  (iter kth-q rest-q)))

In [17]:
(define board (adjoin-position 1 1 empty-board))
board

((1 . 1))

In [18]:
(safe? 1 board)

#t

In [34]:
(safe? 1 (adjoin-position 1 2 board))

#f

In [37]:
(safe? 1 (adjoin-position 3 5 board))

#t

In [38]:
(safe? 1 (adjoin-position 2 7 (adjoin-position 3 5 board)))

#t

In [19]:
(define (queens board-size)
   (define (queen-cols k) 
     (if (= k 0)
         (list empty-board)
         (filter
          (lambda (positions) (safe? k positions))
          (flatmap
           (lambda (rest-of-queens)
             (map (lambda (new-row)
                    (adjoin-position new-row k rest-of-queens))
                  (enumerate-interval 1 board-size)))
           (queen-cols (- k 1))))))
   (queen-cols board-size))

In [40]:
(queens 1)

(((1 . 1)))

In [41]:
(queens 2)

()

In [42]:
(queens 3)

()

In [43]:
(queens 4)

(((2 . 1) (4 . 2) (1 . 3) (3 . 4)) ((3 . 1) (1 . 2) (4 . 3) (2 . 4)))

In [44]:
(queens 5)

(((1 . 1) (3 . 2) (5 . 3) (2 . 4) (4 . 5)) ((1 . 1) (4 . 2) (2 . 3) (5 . 4) (3 . 5)) ((2 . 1) (4 . 2) (1 . 3) (3 . 4) (5 . 5)) ((2 . 1) (5 . 2) (3 . 3) (1 . 4) (4 . 5)) ((3 . 1) (1 . 2) (4 . 3) (2 . 4) (5 . 5)) ((3 . 1) (5 . 2) (2 . 3) (4 . 4) (1 . 5)) ((4 . 1) (1 . 2) (3 . 3) (5 . 4) (2 . 5)) ((4 . 1) (2 . 2) (5 . 3) (3 . 4) (1 . 5)) ((5 . 1) (2 . 2) (4 . 3) (1 . 4) (3 . 5)) ((5 . 1) (3 . 2) (1 . 3) (4 . 4) (2 . 5)))

In [45]:
(queens 6)

(((2 . 1) (4 . 2) (6 . 3) (1 . 4) (3 . 5) (5 . 6)) ((3 . 1) (6 . 2) (2 . 3) (5 . 4) (1 . 5) (4 . 6)) ((4 . 1) (1 . 2) (5 . 3) (2 . 4) (6 . 5) (3 . 6)) ((5 . 1) (3 . 2) (1 . 3) (6 . 4) (4 . 5) (2 . 6)))

## 2.43
Original `queens` `flatmap`section:
```scheme
(flatmap
 (lambda (rest-of-queens)
   (map (lambda (new-row)
          (adjoin-position new-row k rest-of-queens))
        (enumerate-interval 1 board-size)))
 (queen-cols (- k 1)))
```
Louis `queens` `flatmap` section:
```scheme
(flatmap
 (lambda (new-row)
   (map (lambda (rest-of-queens)
          (adjoin-position 
           new-row k rest-of-queens))
        (queen-cols (- k 1))))
 (enumerate-interval 1 board-size))
```

`queen-cols` is called once for each every each column on the board in the original implementation - it's result is used as the `seq` parameter to flatmap.
- **Linear** recursive process

In Louis' implementations, `queen-cols` is called within the `proc` lambda passed into the `flatmap`. The `flatmap` is called once for each row of the `k`th column of the board. This causes **all** the possible solutions for the first `k - 1` columns to be generated for **each row**.
- **Tree** recursive process

Louis implementation will take $T^{b}$ time to execute
- $T$ = time to execute orginal implementation
- $b$ = board size
- Tree recursive processes grow exponentially


## 2.44

In [5]:
(define (up-split painter n)
  (if (= n 0)
      painter
      (let ((smaller (up-split painter
                                  (- n 1))))
        (below painter
               (beside smaller smaller)))))

## 2.45

In [6]:
(define (split direction1 direction2)
  (lambda (painter n)
    (if (= n 0)
        painter
        (let ((smaller ((split direction1 direction2)
                        painter (- n 1))))
          (direction1 painter
                      (direction2 smaller smaller))))))

## 2.46

In [7]:
(define (make-vect x y)
  (cons x y))

(define (xcor-vect v)
  (car v))

(define (ycor-vect v)
  (cdr v))

(define (add-vect v1 v2)
  (make-vect (+ (xcor-vect v1) (xcor-vect v2))
             (+ (ycor-vect v1) (ycor-vect v2))))

(define (sub-vect v1 v2)
  (make-vect (- (xcor-vect v1) (xcor-vect v2))
             (- (ycor-vect v1) (ycor-vect v2))))

(define (scale-vect s v)
  (make-vect (* s (xcor-vect v))
             (* s (ycor-vect v))))

In [8]:
(define v1 (make-vect 1 2))
v1

(1 . 2)

In [9]:
(define v2 (make-vect 5 6))
v2

(5 . 6)

In [10]:
(xcor-vect v1)

1

In [11]:
(ycor-vect v1)

2

In [12]:
(add-vect v1 v2)

(6 . 8)

In [13]:
(sub-vect v1 v2)

(-4 . -4)

In [14]:
(scale-vect 10 v1)

(10 . 20)

## 2.47

In [15]:
(define (make-frame origin edge1 edge2)
  (list origin edge1 edge2))

(define (origin-frame frame)
  (car frame))

(define (edge1-frame frame)
  (car (cdr frame)))

(define (edge2-frame frame)
  (cdr (cdr (cdr frame))))

(define (make-frame origin edge1 edge2)
  (cons origin (cons edge1 edge2)))

(define (origin-frame frame)
  (car frame))

(define (frame-edges frame)
  (cdr frame))

(define (edge1-frame frame)
  (car (frame-edges frame)))

(define (edge2-frame frame)
  (cdr (frame-edges frame)))

## 2.48

In [20]:
(define (make-segment v1 v2)
  (cons v1 v2))

(define (start-segment segment)
  (car segment))

(define (end-segment segment)
  (cdr segment))

In [21]:
(define start-vector (make-vect 1 2))
(define end-vector (make-vect 1 1))
(define s (make-segment start-vector end-vector))
s

((1 . 2) 1 . 1)

In [22]:
(start-segment s)

(1 . 2)

In [23]:
(end-segment s)

(1 . 1)

## 2.49

In [31]:
(define (frame-coord-map frame)
  (lambda (v)
    (add-vect
     (origin-frame frame)
     (add-vect 
      (scale-vect (xcor-vect v)
                  (edge1-frame frame))
      (scale-vect (ycor-vect v)
                  (edge2-frame frame))))))

(define (segments->painter segment-list)
  (lambda (frame)
    (for-each
     (lambda (segment)
       (draw-line
        ((frame-coord-map frame)
         (start-segment segment))
        ((frame-coord-map frame)
         (end-segment segment))))
     segment-list)))

;; would put all inside a `let` to avoid global defines, but calysto scheme
;; will evaluate when cell is run and error due to operations in 
;; segments->painter not being implemented
(define bottom-left (make-vect 0 0))
(define bottom-right (make-vect 1 0))
(define top-left (make-vect 0 1))
(define top-right (make-vect 1 1))
  
(define frame-outline
(segments->painter (list
                    (make-segment bottom-left bottom-right)
                    (make-segment bottom-right top-right)
                    (make-segment top-right top-left)
                    (make-segment top-left bottom-left))))
  
(define corners-x
(segments->painter
 (list 
  (make-segment bottom-left top-right)
  (make-segment bottom-right top-left))))

(define diamond
  (let ((left-mid (make-vect 0 0.5))
        (bottom-mid (make-vect 0.5 0))
        (right-mid (make-vect 1 0.5))
        (top-mid (make-vect 0.5 1)))
  (segments->painter (list
                      (make-segment left-mid top-mid)
                      (make-segment top-mid right-mid)
                      (make-segment right-mid bottom-mid)
                      (make-segment bottom-mid left-mid)))))

`wave` painter would have 17 segments defining the straight edges shown in the image below. Not implemented as large amount of fiddly approximations to get the shape right.
![](images/wave-painter.png)

## 2.50

In [37]:
(define (transform-painter 
         painter origin corner1 corner2)
  (lambda (frame)
    (let ((m (frame-coord-map frame)))
      (let ((new-origin (m origin)))
        (painter (make-frame new-origin
                  (sub-vect (m corner1) 
                            new-origin)
                  (sub-vect (m corner2)
                            new-origin)))))))

(define (rotate90 painter)
  (transform-painter painter
                     (make-vect 1.0 0.0)
                     (make-vect 1.0 1.0)
                     (make-vect 0.0 0.0)))

(define (flip-horiz painter)
  (transform-painter
   painter
   (make-vect 1 0)
   (make-vect 0 0)
   (make-vect 1 1)))

(define (rotate180 painter)
  (transform-painter
   painter
   (make-vect 1 1)
   (make-vect 0 1)
   (make-vect 1 0)))

(define (rotate270 painter)
  (transform-painter
   painter
   (make-vect 0 1)
   (make-vect 0 0)
   (make-vect 1 1)))

## 2.51

In [42]:
(define (beside painter1 painter2)
  (let ((split-point (make-vect 0.5 0.0)))
    (let ((paint-left  (transform-painter 
                        painter1
                        (make-vect 0.0 0.0)
                        split-point
                        (make-vect 0.0 1.0)))
          (paint-right (transform-painter
                        painter2
                        split-point
                        (make-vect 1.0 0.0)
                        (make-vect 0.5 1.0))))
      (lambda (frame)
        (paint-left frame)
        (paint-right frame)))))

;; `below` draws with painter1 in bottom of frame and painter2 in top
;; method 1 - analagous to `beside`
(define (below painter1 painter2)
  (let ((split-point (make-vect 0 0.5)))
    (let ((paint-bottom (transform-painter
                         painter1
                         (make-vect 0 0)
                         (make-vect 1 0)
                         split-point))
          (paint-top (transform-painter
                      painter2
                      split-point
                      (make-vect 1 0.5)
                      (make-vect 0 1))))
    (lambda (frame)
      (paint-bottom frame)
      (paint-top frame)))))

;;method 2 - in terms of `beside` and rotation operations
;; Get painters to draw on top of each other:
;;  -> draw painters beside each other, then rotate frame 90 deg
;; Re-orientate images in each section to be correct way up:
;;  -> rotate each painter 270 deg
(define (below painter1 painter2)
  (rotate90 (beside (rotate270 painter1) (rotate270 painter2))))

## 2.52
N.B Haven't implemented question 1 due to not fully implementing `wave` in exercise 2.49. However, adding a smile would involce adding several new segments to the list passed into `segments->painter`
- i.e. 3 small segments that join to form a crude square smile in the middle of the face:
    ```
    \__/
    ```

In [44]:
;; use only one copy of up-split and right-split images instead of two
(define (corner-split painter n)
  (if (= n 0)
      painter
      (let ((up (up-split painter (- n 1)))
            (right (right-split painter 
                                (- n 1)))
              (corner (corner-split painter 
                                    (- n 1))))
          (beside (below painter up)
                  (below right 
                         corner)))))

(define (square-of-four tl tr bl br)
  (lambda (painter)
    (let ((top (beside (tl painter) 
                       (tr painter)))
          (bottom (beside (bl painter) 
                          (br painter))))
      (below bottom top))))

;; 'Mr Rogers' image would look outwards from each corner of the square
(define (square-limit painter n)
  (let ((combine4 
         (square-of-four flip-vert 
                         rotate180
                         identitiy 
                         flip-horiz)))
    (combine4 (corner-split painter n))))

## 2.53
```scheme
(a b c)
((george))
((y1 y2))
(y1 y2)
#f
#f
(red shoes blue socks)
```

In [47]:
(list 'a 'b 'c)

(a b c)

In [48]:
(list (list 'george))

((george))

In [50]:
(cdr '((x1 x2) (y1 y2)))

((y1 y2))

In [51]:
(cadr '((x1 x2) (y1 y2)))

(y1 y2)

In [52]:
(pair? (car '(a short list)))

#f

In [53]:
(memq 'red '((red shoes) (blue socks)))

#f

In [54]:
(memq 'red '(red shoes blue socks))

(red shoes blue socks)

## 2.54

In [67]:
(define (equal? x y)
  (cond ((and (null? x) (null? y)) #t)
        ((or (null? x) (null? y)) #f) ;; null != any symbol other than null
        ((and (pair? x) (pair? y)) ;; both lists
         (and (equal? (car x) (car y)) ;; check lists equal
              (equal? (cdr x) (cdr y))))
        ((or (pair? x) (pair? y)) #f) ;; one list, one symbol
        (else (eq? x y)))) ;; both symbols

In [68]:
(equal? '() '())

#t

In [69]:
(equal? '() 'x)

#f

In [70]:
(equal? '(x y z) '(x y z))

#t

In [71]:
(equal? '(x y z) '(x y))

#f

In [72]:
(equal? '(x y z) 'x)

#f

In [73]:
(equal? 'x 'x)

#t

In [74]:
(equal? 1 1)

#t

## 2.55
As `'` is shorthand for `quote` `''abracadabra` is equivalent to `(quote (quote abracadabra))`:

In [85]:
;; note additional ' at start to show symbols
'''abracadabra

(quote (quote abracadabra))

`(quote (quote abracadabra))` evaluates to a list of symbols `(quote abracadabra)`

In [86]:
(quote (quote abracadabra))

(quote abracadabra)

`car` of the list returns `quote`

In [88]:
(car (quote (quote abracadabra)))

quote

## Algebraic Expression Procedures for 2.56-2.58

In [133]:
(define (deriv exp var)
  (cond ((number? exp) 0)
        ((variable? exp)
         (if (same-variable? exp var) 1 0))
        ((sum? exp)
         (make-sum (deriv (addend exp) var)
                   (deriv (augend exp) var)))
        ((product? exp)
         (make-sum
          (make-product
           (multiplier exp)
           (deriv (multiplicand exp) var))
          (make-product
           (deriv (multiplier exp) var)
           (multiplicand exp))))
        (else (error "unknown expression
                     type: DERIV" exp))))

(define (variable? x) (symbol? x))

(define (=number? exp num)
  (and (number? exp) (= exp num)))
(define (same-variable? v1 v2)
  (and (variable? v1)
       (variable? v2)
       (eq? v1 v2)))

(define (make-sum a1 a2)
  (cond ((=number? a1 0) a2)
        ((=number? a2 0) a1)
        ((and (number? a1) (number? a2)) 
         (+ a1 a2))
        (else (list '+ a1 a2))))

(define (addend s) (cadr s))

(define (augend s) (caddr s))

(define (sum? x)
  (and (pair? x) (eq? (car x) '+)))

(define (make-product m1 m2)
  (cond ((or (=number? m1 0) 
             (=number? m2 0)) 
         0)
        ((=number? m1 1) m2)
        ((=number? m2 1) m1)
        ((and (number? m1) (number? m2)) 
         (* m1 m2))
        (else (list '* m1 m2))))

(define (multiplier p) (cadr p))

(define (multiplicand p) (caddr p))

(define (product? x)
  (and (pair? x) (eq? (car x) '*)))

### Demonstrate Procedures

In [109]:
(variable? 1)

#f

In [110]:
(variable? 'x)

#t

In [111]:
(same-variable? 'x 'y)

#f

In [112]:
(same-variable? 'x 'x)

#t

In [118]:
(define s (make-sum 'x 2))
s

(+ x 2)

In [119]:
(addend s)

x

In [120]:
(augend s)

2

In [121]:
(sum? s)

#t

In [106]:
(product? s)

#f

In [129]:
;; simplify
(make-sum 1 2)

3

In [122]:
(define p (make-product 'x 2))
p

(* x 2)

In [123]:
(multiplier p)

x

In [124]:
(multiplicand p)

2

In [125]:
(product? p)

#t

In [126]:
(sum? p)

#f

In [128]:
;; simplify
(make-product 1 2)

2

In [130]:
(deriv '(+ x 3) 'x)

1

In [132]:
(deriv '(* x y) 'x)

y

In [131]:
(deriv '(* (* x y) (+ x 3)) 'x)

(+ (* x y) (* y (+ x 3)))

## 2.56

In [153]:
(define (deriv exp var)
  (cond ((number? exp) 0)
        ((variable? exp)
         (if (same-variable? exp var) 1 0))
        ((sum? exp)
         (make-sum (deriv (addend exp) var)
                   (deriv (augend exp) var)))
        ((product? exp)
         (make-sum
          (make-product
           (multiplier exp)
           (deriv (multiplicand exp) var))
          (make-product
           (deriv (multiplier exp) var)
           (multiplicand exp))))
        ((exponentiation? exp)
         (make-product 
          (make-product (exponent exp)
                        (make-exponentiation (base exp)
                                             (- (exponent exp) 1)))
          (deriv (base exp) var)))
        (else (error "unknown expression
                     type: DERIV" exp))))

(define (exponentiation? x)
  (and (pair? x) (eq? (car x) '**)))

(define (make-exponentiation base exponent)
  (cond ((=number? exponent 0) 1)
        ((=number? exponent 1) base)
        ((and (number? base) (number? exponent))
         (expt base exponent))
        (else (list '** base exponent))))

(define (base e)
  (cadr e))

(define (exponent e)
  (caddr e))

### Test Exponentiation Procedures

In [155]:
(define e (make-exponentiation 'x 2))
e

(** x 2)

In [156]:
(base e)

x

In [157]:
(exponent e)

2

In [158]:
(exponentiation? e)

#t

In [159]:
(sum? e)

#f

In [160]:
(product? e)

#f

In [164]:
;; simplifications
(make-exponentiation 2 4)

16.0

In [165]:
(make-exponentiation 'x 1)

x

In [166]:
(make-exponentiation 'x 0)

1

### Test `deriv` with new exponent rule

In [172]:
;; x^2
(deriv '(** x 2) 'x)

(* 2 x)

Example from introduction of section 2.3.2:

"if the arguments to the procedure are $ax^2+bx+c$ and $x$, the procedure should return $2ax+b$."

In [173]:
(deriv '(+ (+ (* a (** x 2)) (* b x)) c) 'x)

(+ (* a (* 2 x)) b)

## 2.57

In [191]:
;; change augend and multiplicand to return sum/product of 
;; all but first terms?
(define (augend s) 
  (if (null? (cdddr s))
      (caddr s)
      (cons '+ (cddr s))))

(define s '(+ x 3 y))
s

(+ x 3 y)

In [192]:
(addend s)

x

In [193]:
(augend s)

(+ 3 y)

In [194]:
(define (multiplicand p)
  (if (null? (cdddr p))
      (caddr p)
      (cons '* (cddr p))))

(define p '(* x 3 y))
p

(* x 3 y)

In [195]:
(multiplier p)

x

In [196]:
(multiplicand p)

(* 3 y)

In [197]:
(deriv '(* x y (+ x 3)) 'x)

(+ (* x y) (* y (+ x 3)))

## 2.58

### Infix `sum` representation

In [208]:
(define (make-sum a1 a2)
  (cond ((=number? a1 0) a2)
        ((=number? a2 0) a1)
        ((and (number? a1) (number? a2)) 
         (+ a1 a2))
        (else (list a1 '+ a2))))

(define (addend s) (car s))

(define (augend s) (caddr s))

(define (sum? x)
  (and (pair? x) (eq? (cadr x) '+)))

In [210]:
(define s (make-sum 'x 4))
s

(x + 4)

In [211]:
(addend s)

x

In [212]:
(augend s)

4

In [213]:
(sum? s)

#t

### Infix `product` representation

In [214]:
(define (make-product m1 m2)
  (cond ((or (=number? m1 0)
             (=number? m2 0))
         0)
        ((=number? m1 1) m2)
        ((=number? m2 1) m1)
        ((and (number? m1) (number? m2))
         (* m1 m2))
        (else (list m1 '* m2))))

(define (multiplier p) (car p))

(define (multiplicand p) (caddr p))

(define (product? x)
  (and (pair? x) (eq? (cadr x) '*)))

In [216]:
(define p (make-product 'x 4))
p

(x * 4)

In [217]:
(multiplier p)

x

In [218]:
(multiplicand p)

4

In [219]:
(product? p)

#t

### Test `deriv`
Example from book $x+3(x+(y+2))$

In [222]:
(deriv '(x + (3 * (x + (y + 2)))) 'x)

4

### Allowing for Standard Algebraic Notation
Drop unnecessary parentheses and assume multiplication before addition: $x + 3(x+y+2)$
- `(deriv (x + 3 * (x + y + 2)) x)`

**N.B** This implementation has very little/no error handling and assumes well formed inputs.

The 'type' of an expression is determined by the presence (in order of precedence) of a `+`, `*` or `**` operator anywhere in the expression:
```scheme
'(x + 3) -> sum
'(x * 2) -> product
'(x ** 2) -> exponent

'(x * x + 2) -> sum
'(x ** 2 + 2) -> sum
'(x * x ** 2) -> product
```

`addend`, `multiplier` and `base` selectors should return the first part of the expression passed in up to (not including) the first appearance of `+`, `*` and `**` operators respectively:
```scheme
(addend `((x + 3) + 'y)) -> (x + 3)
```

`augend`, `multiplicand` and `exponent` selectors should return the remaining part of the expression passed in after (not including) the first appearance of the `+`, `*` and `**` operators resectively:
```scheme
(augend `('y + (x + 3))) -> (x + 3)
```

Selectors should unwrap single values so they aren't contained within a list:
```scheme
(addend `(x + 3)) -> x
```

In [261]:
;; returns first operation in expression
(define (operation expr)
  (cond ((memq '+ expr) '+)
        ((memq '* expr) '*)
        ((memq '** expr) '**)))

;; returns section of x before first appearance of symbol
;; will be wrapped in a list
(define (prefix symbol x)
  (if (or (null? x) (eq? symbol (car x)))
      '() 
      (cons (car x) 
            (prefix symbol (cdr x)))))

(define (addend s)
  (let ((a (prefix '+ s)))
    (if (= (length a) 1)
        (car a)
        a)))

(define (augend s)
  (let ((a (cdr (memq '+ s))))
    (if (= (length a) 1)
        (car a)
        a)))

(define (sum? x)
  (eq? '+ (operation x)))

(define (multiplier p)
  (let ((m (prefix '* p)))
    (if (= (length m) 1)
        (car m)
        m)))

(define (multiplicand p)
  (let ((m (cdr (memq '* p))))
    (if (= (length m) 1)
        (car m)
        m)))

(define (product? x)
  (eq? '* (operation x)))


(define (base e)
  (let ((p (prefix '** e)))
    (if (= (length p) 1)
        (car p)
        p)))

(define (exponent e)
(let ((p (cdr (memq '** e))))
  (if (= (length p) 1)
      (car p)
      p)))

(define (exponentiation? x)
  (eq? '** (operation x)))

In [262]:
(deriv '(x + 3) 'x)

1

In [263]:
(deriv '(x * y) 'x)

y

In [264]:
(deriv '(x * y * (x + 3)) 'x)

((x * y) + (y * (x + 3)))

In [265]:
(deriv '(x * 3 + 5 * (x + y + 2)) 'x)

8

## 2.59

In [270]:
(define (element-of-set? x set)
  (cond ((null? set) #f)
        ((equal? x (car set)) #t)
        (else (element-of-set? x (cdr set)))))

(define (adjoin-set x set)
  (if (element-of-set? x set)
      set
      (cons x set)))

(define (intersection-set set1 set2)
   (cond ((or (null? set1) (null? set2))
          '())
         ((element-of-set? (car set1) set2)
          (cons (car set1)
                (intersection-set (cdr set1)
                                  set2)))
         (else (intersection-set (cdr set1)
                                 set2))))

In [271]:
(element-of-set? 1 (list 1 2 3))

#t

In [276]:
(element-of-set? 4 (list 1 2 3))

#f

In [274]:
(adjoin-set 4 (list 1 2 3))

(4 1 2 3)

In [275]:
(adjoin-set 1 (list 1 2 3))

(1 2 3)

In [277]:
(intersection-set (list 1 2 3) (list 1 2 3))

(1 2 3)

In [278]:
(intersection-set (list 1 2 3) (list 1 2))

(1 2)

In [280]:
(intersection-set (list 1 2 3) (list 4 5 6))

()

In [282]:
(define (union-set set1 set2)
  (cond ((null? set1) set2)
        ((null? set2) set1)
        ((element-of-set? (car set1) set2)
         (union-set (cdr set1) set2))
        (else (cons (car set1) 
                    (union-set (cdr set1) set2)))))

In [283]:
(union-set (list 1 2 3) (list 4 5 6))

(1 2 3 4 5 6)

In [284]:
(union-set (list 1 2 3) (list 1 2 3))

(1 2 3)

In [285]:
(union-set (list 1 2 3) (list 3 4 5))

(1 2 3 4 5)

In [286]:
(union-set (list 1 2 3) '())

(1 2 3)

## 2.60
`element-of-set?` requires no change

In [ ]:
(define (adjoin-set x set)
  (cons x set))

(define (union-set set1 set2)
  (append set1 set2))

Question is ambiguous - stating that the representation should *allow* duplicates, not that is *must* maintain all duplicates. This means that `intersection-set` can remain unchanged depending on interpretation.

The current implementation checks the elements of the first set to determine if they appear in the second set.
- Duplicates will only appear in the result if the first set contains them

This doesn't affect the properties of the `set` as the correct elements will still be present, however the *number of duplicates* isn't necessarily maintained:

In [288]:
(intersection-set (list 2 2 2) (list 2 3 4))

(2 2 2)

In [289]:
(intersection-set (list 2 3 4) (list 2 2 2))

(2)

Since the resulting 'set' is equivalent, I'm leaving `intersection-set` unchanged.

### Efficiency
`element-of-set?` and `intersection-set` are unchanged -> performance unchanged.

`adjoin-set` and `union-set` have improved as they no longer check for duplicate elements
- `adjoin-set` = $O(1)$
- `union-set` = $O(n)$

Overall **memory usage** has increased, as the `list` representing the set will now contain duplicate elements and can grow to be much larger whilst representing the same set.

If the application is not memory constrained and a large number of `adjoin-set` or `union-set` operations are being performed, this representation is preferable. 

If the application is memory constrained and a large number of `intersection-set` and `element-of-set?` are being performed then the original representation is preferable.

## 2.61
Set representation using **ordered** lists.

In [335]:
(define (element-of-set? x set)
  (cond ((null? set) #f)
        ((= x (car set)) #t)
        ((< x (car set)) #f)
        (else (element-of-set? x (cdr set)))))

(define (intersection-set set1 set2)
  (if (or (null? set1) (null? set2))
      '()
      (let ((x1 (car set1)) (x2 (car set2)))
        (cond ((= x1 x2)
               (cons x1 (intersection-set 
                         (cdr set1)
                         (cdr set2))))
              ;; x2 is smallest element in set2 -> x1 not in set2
              ;; -> x1 not in intersection
              ((< x1 x2) (intersection-set 
                          (cdr set1) 
                          set2))
              ;; x1 is smallest element in set1 -> x2 not in set1
              ;; -> x2 not in intersection
              ((< x2 x1) (intersection-set 
                          set1 
                          (cdr set2)))))))

In [343]:
;; need to insert x into correct position to maintain order
;; if x is encountered in the set -> return the set unchanged
(define (adjoin-set x set)
   (cond ((null? set) (list x))
         ((= x (car set)) set)
         ((< x (car set)) (cons x set))
         ((> x (car set)) (cons (car set)
                                (adjoin-set x (cdr set))))))

In [344]:
(adjoin-set 1 '())

(3)

In [345]:
(adjoin-set 1 '(2 3 4))

(1 2 3 4)

In [347]:
(adjoin-set 5 '(2 3 4))

(2 3 4 5)

In [348]:
(adjoin-set 5 '(2 3 4 6))

(2 3 4 5 6)

In [349]:
(adjoin-set 2 '(2 3 4))

(2 3 4)

## 2.62

In [356]:
;; previous union-set for reference
(define (union-set-unordered set1 set2)
  (cond ((null? set1) set2)
        ((null? set2) set1)
        ((element-of-set? (car set1) set2)
         (union-set (cdr set1) set2))
        (else (cons (car set1) 
                    (union-set (cdr set1) set2)))))

;; return list of elements in both sets, ordered, without dups
;; both sets ordered -> join by comparing element-wise
;; -> decide which to place in resulting set
;; elements equal -> place either into set, call union-set with cdr of both
;; set1 element < set2 element -> place set1 element into set,
;; call union-set with cdr set1 and whole of set2
;; else place set2 element into set, 
;; call union set with cdr set2 and whole set1
(define (union-set set1 set2)
  (cond ((null? set1) set2)
        ((null? set2) set1)
        (else
         (let ((x1 (car set1)) (x2 (car set2)))
           (cond ((= x1 x2)
                  (cons x1 (union-set (cdr set1) (cdr set2))))
                 ((< x1 x2)
                  (cons x1 (union-set (cdr set1) set2)))
                 (else (cons x2 (union-set set1 (cdr set2)))))))))

In [357]:
(union-set (list 1 2 3) (list 1 2 3))

(1 2 3)

In [358]:
(union-set (list 1 2 3) (list 4 5 6))

(1 2 3 4 5 6)

In [359]:
(union-set (list 4 5 6) (list 1 2 3))

(1 2 3 4 5 6)

In [360]:
(union-set (list 1 2 3) '())

(1 2 3)

## Sets as Binary Trees

In [3]:
(define (entry tree) (car tree))
(define (left-branch tree) (cadr tree))
(define (right-branch tree) (caddr tree))
(define (make-tree entry left right)
  (list entry left right))

(define (element-of-set? x set)
  (cond ((null? set) #f)
        ((= x (entry set)) #t)
        ((< x (entry set))
         (element-of-set?
          x
          (left-branch set)))
        ((> x (entry set))
         (element-of-set?
          x
          (right-branch set)))))

(define (adjoin-set x set)
  (cond ((null? set) (make-tree x '() '()))
        ((= x (entry set)) set)
        ((< x (entry set))
         (make-tree
          (entry set)
          (adjoin-set x (left-branch set))
          (right-branch set)))
        ((> x (entry set))
         (make-tree
          (entry set)
          (left-branch set)
          (adjoin-set x (right-branch set))))))

In [4]:
(define t (make-tree 7 '() '()))
t

(7 () ())

In [363]:
(entry t)

7

In [364]:
(left-branch t)

()

In [365]:
(right-branch t)

()

In [366]:
(element-of-set? 7 t)

#t

In [373]:
(define t (adjoin-set 8 t))
t

(7 () (8 () ()))

In [374]:
(entry t)

7

In [375]:
(left-branch t)

()

In [376]:
(right-branch t)

(8 () ())

In [377]:
(element-of-set? 8 t)

#t

In [378]:
(define t (adjoin-set 5 t))
t

(7 (5 () ()) (8 () ()))

In [379]:
(entry t)

7

In [380]:
(left-branch t)

(5 () ())

In [381]:
(right-branch t)

(8 () ())

## 2.63

In [5]:
(define (tree->list-1 tree)
  (if (null? tree)
      '()
      (append
       (tree->list-1
        (left-branch tree))
       (cons (entry tree)
             (tree->list-1
              (right-branch tree))))))

(define (tree->list-2 tree)
  (define (copy-to-list tree result-list)
    (if (null? tree)
        result-list
        (copy-to-list
         (left-branch tree)
         (cons (entry tree)
               (copy-to-list
                (right-branch tree)
                result-list)))))
  (copy-to-list tree '()))

;; trees from fig 2.16
(define t1 
  (make-tree 7 
             (make-tree 3 
                        (make-tree 1 '() '()) 
                        (make-tree 5 '() '()))
             (make-tree 9
                        '()
                        (make-tree 11 '() '()))))

(define t2
  (make-tree 3
             (make-tree 1 '() '())
             (make-tree 7
                        (make-tree 5 '() '())
                        (make-tree 9 
                                   '()
                                   (make-tree 11 
                                              '() 
                                              '())))))

(define t3
  (make-tree 5
             (make-tree 3
                        (make-tree 1 '() '())
                        '())
             (make-tree 9
                        (make-tree 7 '() '())
                        (make-tree 11 '() '()))))

`tree->list-1` and `tree->list-2` produce the same results for every tree - performing an **in order** traversal.

In [390]:
(tree->list-1 t1)

(1 3 5 7 9 11)

In [391]:
(tree->list-2 t1)

(1 3 5 7 9 11)

In [392]:
(tree->list-1 t2)

(1 3 5 7 9 11)

In [393]:
(tree->list-2 t2)

(1 3 5 7 9 11)

In [394]:
(tree->list-1 t3)

(1 3 5 7 9 11)

In [395]:
(tree->list-2 t3)

(1 3 5 7 9 11)

`tree->list-1` calls the `append` procedure in each recursive call. `append` has $O(n)$ order of growth where $n$ is the length of the **first list** passed in:
```scheme
(define (append list1 list2)
  (if (null? list1)
      list2
      (cons (car list1) (append (cdr list1) list2))))
```

The first list passed to `append` in `tree->list-1` is the left branch of the tree. For a *balanced* tree, the left branch will contain half the nodes of the tree. The number of nodes in the left branch is **halved** in each recursive call to `tree->list-1` -> $O(logn)$ list size. This is done for each $n$ calls to `append` so `tree->list-1` has **$O(nlogn)$ order of growth for a balanced tree**.

`tree->list-2` calls `cons` for each recursive call. `cons` is a **constant time** operation with $O(1)$ order of growth. Therefore, the order of growth for `tree->list-2` is directly proportional to the size of the tree - $O(n)$

## 2.64

In [6]:
(define (list->tree elements)
  (car (partial-tree
        elements (length elements))))

;; construct balanced tree containing first `n` elements of `elts`
;; returns pair of form (tree elts-not-in-tree)
(define (partial-tree elts n)
  (if (= n 0)
      (cons '() elts)
      (let ((left-size
             (quotient (- n 1) 2)))
        (let ((left-result
               (partial-tree
                elts left-size)))
          (let ((left-tree
                 (car left-result))
                (non-left-elts
                 (cdr left-result))
                (right-size
                 (- n (+ left-size 1))))
            (let ((this-entry
                   (car non-left-elts))
                  (right-result
                   (partial-tree
                    (cdr non-left-elts)
                    right-size)))
              (let ((right-tree
                    (car right-result))
                    (remaining-elts
                     (cdr right-result)))
              (cons (make-tree this-entry
                               left-tree
                               right-tree)
                    remaining-elts))))))))

`partial-tree` recursively constructs left and right halves of the sub-trees within the final balanced tree. In each call, it 'splits' `elts` around the center element. Elements to the left of the center are passed to a recursive call to `partial-tree` to construct the left branch, elements to the right of the center are passed to a recursive call to `partial-tree` to construct the right branch and the center element becomes the root node of the tree. This continues until `elts` is empty (no elements remaining to be added) and the sub-trees are assembled into the full balanced tree.

`list->tree` produces the following tree for the list `(1 3 5 7 9 11)`:
```
    5
   / \
  /   \
 1     9
  \   / \
   3 7   11
```

In [403]:
(list->tree (list 1 3 5 7 9 11))

(5 (1 () (3 () ())) (9 (7 () ()) (11 () ())))

In total `list->tree` produces a single call to `partial-tree` for every element in the input list. The work performed in `partial-tree` is a `cons` operation which has $O(1)$ 

## 2.65

In [7]:
;; union-set for ordered lists from 2.62
(define (union-list list1 list2)
  (cond ((null? list1) list2)
        ((null? list2) list1)
        (else
         (let ((x1 (car list1)) (x2 (car list2)))
           (cond ((= x1 x2)
                  (cons x1 (union-set (cdr list1) (cdr list2))))
                 ((< x1 x2)
                  (cons x1 (union-set (cdr list1) list2)))
                 (else (cons x2 (union-set list1 (cdr list2)))))))))

(define (union-set tree1 tree2)
  (list->tree (union-list (tree->list-2 tree1)
                          (tree->list-2 tree 2))))

;; intersection-set for ordered lists from 2.61
(define (intersection-list list1 list2)
  (if (or (null? list1) (null? list2))
      '()
      (let ((x1 (car list1)) (x2 (car list2)))
        (cond ((= x1 x2)
               (cons x1 (intersection-set 
                         (cdr list1)
                         (cdr list2))))
              ;; x2 is smallest element in list2 -> x1 not in list2
              ;; -> x1 not in intersection
              ((< x1 x2) (intersection-set 
                          (cdr list1) 
                          list2))
              ;; x1 is smallest element in list1 -> x2 not in list1
              ;; -> x2 not in intersection
              ((< x2 x1) (intersection-set 
                          list1 
                          (cdr list2)))))))

(define (intersection-set tree1 tree2)
  (list->tree (intersection-list (tree->list-2 tree1)
                                 (tree->list-2 tree 2))))

## 2.66

Need constructors/selectors for records:
- `make-record`
- `record-key`
- `record-data`

Implement `lookup` based on `element-of-set?` implementation for binary tree set representation
- Modify to return record if found instead of `#t`

Build initial database as a list of records then convert to tree using `list->tree`

In [8]:
(define (make-record key data)
  (cons key data))

(define (record-key record)
  (car record))

(define (record-data record)
  (cdr record))

In [16]:
(define r (make-record 1 "Important Information"))
r

(1 . "Important Information")

In [17]:
(record-key r)

1

In [18]:
(record-data r)

"Important Information"

In [9]:
(define (element-of-set? x set)
  (cond ((null? set) #f)
        ((= x (entry set)) #t)
        ((< x (entry set))
         (element-of-set?
          x
          (left-branch set)))
        ((> x (entry set))
         (element-of-set?
          x
          (right-branch set)))))

(define (lookup given-key set-of-records)
  (cond ((null? set-of-records) #f)
        ((= given-key (record-key (car set-of-records))) (car set-of-records))
        ((< given-key (record-key (car set-of-records)))
         (lookup
          given-key
          (left-branch set-of-records)))
        ((> given-key (record-key (car set-of-records)))
         (lookup
          given-key
          (right-branch set-of-records)))))

In [43]:
(define database-list
  (list
   (make-record 1 "data 1")
   (make-record 2 "data 2")
   (make-record 3 "data 3")
   (make-record 10 "data 10")))
database-list

((1 . "data 1") (2 . "data 2") (3 . "data 3") (10 . "data 10"))

In [44]:
(define database-tree (list->tree database-list))
database-tree

((2 . "data 2") ((1 . "data 1") () ()) ((3 . "data 3") () ((10 . "data 10") () ())))

In [45]:
(lookup 1 database-tree)

(1 . "data 1")

In [46]:
(lookup 10 database-tree)

(10 . "data 10")

In [47]:
(lookup 100 database-tree)

#f

# Huffman Encoding
**Leaves** of tree represented by a list of the *symbol* `leaf`, the symbol at the leaf and the *weight*.

General tree represented as a list of a left branch, right branch, set of symbols and a weight
- Set of symbols = list of symbols

Make a tree by merging two nodes
- Weight of tree = sum of node weights
- Set of symbols = union of node symbols

Decoding:
- Input message as list of bits and tree
- Move down tree according to value of next bit
    - `0` = left branch
    - `1` = right branch
- When leaf node reached, return symbol at leaf as next symbol in the message
    - `cons` with result of decoding rest of message

In [13]:
;; append procedure from 2.18
(define (append list1 list2)
  (if (null? list1)
      list2
      (cons (car list1)
            (append (cdr list1)
                    list2))))

(define (make-leaf symbol weight)
  (list 'leaf symbol weight))

(define (leaf? object)
  (eq? (car object) 'leaf))

(define (symbol-leaf x) (cadr x))

(define (weight-leaf x) (caddr x))

(define (make-code-tree left right)
  (list left
        right
        (append (symbols left)
                (symbols right))
        (+ (weight left) (weight right))))

(define (left-branch tree) (car tree))
(define (right-branch tree) (cadr tree))

(define (symbols tree)
  (if (leaf? tree)
      (list (symbol-leaf tree))
      (caddr tree)))

(define (weight tree)
  (if (leaf? tree)
      (weight-leaf tree)
      (cadddr tree)))

In [16]:
(define l (make-leaf 'x 10))
l

(leaf x 10)

In [17]:
(leaf? l)

#t

In [18]:
(leaf? (list 'not-leaf 'x 10))

#f

In [19]:
(symbol-leaf l)

x

In [20]:
(weight-leaf l)

10

In [24]:
(define l-tree
  (make-code-tree
   l
   (make-leaf 'y 5)))

(define r-tree
  (make-code-tree
   (make-leaf 'a 1)
   (make-leaf 'b 2)))

(define t
  (make-code-tree l-tree r-tree))

In [25]:
l-tree

((leaf x 10) (leaf y 5) (x y) 15)

In [28]:
(symbols l-tree)

(x y)

In [29]:
(weight l-tree)

15

In [30]:
(left-branch l-tree)

(leaf x 10)

In [31]:
(right-branch l-tree)

(leaf y 5)

In [32]:
r-tree

((leaf a 1) (leaf b 2) (a b) 3)

In [27]:
t

(((leaf x 10) (leaf y 5) (x y) 15) ((leaf a 1) (leaf b 2) (a b) 3) (x y a b) 18)

In [33]:
(symbols t)

(x y a b)

In [34]:
(weight t)

18

In [35]:
(left-branch t)

((leaf x 10) (leaf y 5) (x y) 15)

In [36]:
(right-branch t)

((leaf a 1) (leaf b 2) (a b) 3)

In [37]:
(define (decode bits tree)
  (define (decode-1 bits current-branch)
    (if (null? bits)
        '()
        (let ((next-branch
               (choose-branch
                (car bits)
                current-branch)))
          (if (leaf? next-branch)
              (cons
               (symbol-leaf next-branch)
               (decode-1 (cdr bits) tree))
              (decode-1 (cdr bits)
                        next-branch)))))
  (decode-1 bits tree))

(define (choose-branch bit branch)
  (cond ((= bit 0) (left-branch branch))
        ((= bit 1) (right-branch branch))
        (else (error "bad bit:
                     CHOOSE-BRANCH" bit))))

In [39]:
;; add x to set, comparing items by weight
;; assume x is never already in set
(define (adjoin-set x set)
  (cond ((null? set) (list x))
        ((< (weight x) (weight (car set)))
         (cons x set))
        (else
         (cons (car set)
               (adjoin-set x (cdr set))))))

;; take list of symbol-frequency pairs i.e ((A 4) (B 2) (C 1) (D 1))
;; and constructs initial ordered set of leaves to be merged according 
;; to Huffman algo
(define (make-leaf-set pairs)
  (if (null? pairs)
      '()
      (let ((pair (car pairs)))
        (adjoin-set
         (make-leaf (car pair)
                    (cadr pair))
         (make-leaf-set (cdr pairs))))))

## 2.67

In [56]:
(define sample-tree
  (make-code-tree
   (make-leaf 'A 4)
   (make-code-tree
    (make-leaf 'B 2)
    (make-code-tree
     (make-leaf 'D 1)
     (make-leaf 'C 1)))))

(define sample-message
  '(0 1 1 0 0 1 0 1 0 1 1 1 0))

(decode sample-message sample-tree)

(A D A B B C A)

## 2.68
`encode-symbol` - return list of bits encoding symbol according to tree
- Traverse tree
- If tree is leaf -> return empty list
- If symbol in set of symbols of current left branch:
    - Add 0 to result and recurse on left branch
- If symbol in set of symbols of current right branch:
   - Add 1 to result and recurse on right branch
- Else signal error
    - Symbol not in tree

In [79]:
;; original element-of-set? from 2.59
(define (element-of-set? x set)
  (cond ((null? set) #f)
        ((eq? x (car set)) #t)
        (else (element-of-set? x (cdr set)))))

(define (encode message tree)
  (if (null? message)
      '()
      (append
       (encode-symbol (car message)
                      tree)
       (encode (cdr message) tree))))

(define (encode-symbol symbol tree)
  (if (leaf? tree)
      '()
      (let ((l-branch (left-branch tree))
            (r-branch (right-branch tree)))
        (cond ((element-of-set? symbol (symbols l-branch))
               (cons 0 (encode-symbol symbol l-branch)))
              ((element-of-set? symbol (symbols r-branch))
               (cons 1 (encode-symbol symbol r-branch)))
              (else (error "symbol not in tree:
                           ENCODE-SYMBOL" symbol))))))

In [80]:
(define decoded-message (decode sample-message sample-tree))
decoded-message

(A D A B B C A)

In [81]:
(define result (encode decoded-message sample-tree))
result

(0 1 1 0 0 1 0 1 0 1 1 1 0)

In [82]:
(equal? result sample-message)

#t

## 2.69

In [84]:
(define (generate-huffman-tree pairs)
  (successive-merge
   (make-leaf-set pairs)))

In [113]:
(define (successive-merge set)
  (cond ((null? set) '())
        ((= (length set) 1) (car set))
        (else (successive-merge 
               (adjoin-set 
                (make-code-tree (car set) (cadr set)) 
                (cddr set))))))

(generate-huffman-tree '((A 4) (B 2) (C 1) (D 1)))

((leaf A 4) ((leaf B 2) ((leaf D 1) (leaf C 1) (D C) 2) (B D C) 4) (A B D C) 8)

## 2.70

In [115]:
(define pairs '((A 2) (NA 16) (BOOM 1) (SHA 3) (GET 2) (YIP 9) (JOB 2) (WAH 1)))
(define tree (generate-huffman-tree pairs))
tree

((leaf NA 16) ((leaf YIP 9) (((leaf A 2) ((leaf WAH 1) (leaf BOOM 1) (WAH BOOM) 2) (A WAH BOOM) 4) ((leaf SHA 3) ((leaf JOB 2) (leaf GET 2) (JOB GET) 4) (SHA JOB GET) 7) (A WAH BOOM SHA JOB GET) 11) (YIP A WAH BOOM SHA JOB GET) 20) (NA YIP A WAH BOOM SHA JOB GET) 36)

In [117]:
(define message '(GET A JOB SHA NA NA NA NA NA NA NA NA GET A JOB SHA NA
                      NA NA NA NA NA NA NA WAH YIP YIP YIP YIP YIP YIP 
                      YIP YIP YIP SHA BOOM))

(define encoded (encode message tree))
encoded

(1 1 1 1 1 1 1 0 0 1 1 1 1 0 1 1 1 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 0 0 1 1 1 1 0 1 1 1 0 0 0 0 0 0 0 0 0 1 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 1 1 0 1 1 0 1 1)

In [118]:
;; bits required for encoding
(length encoded)

84

Encoding with fixed length requires $log_2(n)$ bits per token where $n$ is the number of symbols in the alphabet.

8 symbols alphabet = $log_2(8)=3$ bits per token

Message is 36 tokens -> bits required = $36*3=108$ 

## 2.71
Alphabet of $n$ symbols, where relative frequencies are $1,2,4,...2^n-1$

### $n=5$
$\{A,B,C,D,E\}$

Frequencies = $1,2,4,8,16$
```
                    {A,B,C,D,E} 31
                        / \
            {A,B,C,D} 15  {E} 16
                / \
        {A,B,C} 7  {D} 8
          / \
    {A,B} 3  {C} 4
     / \
{A} 1   {B} 2
```
### $n=10$
$\{A,B,C,D,E,F,G,H,I,J\}$

Frequencies = $1,2,4,8,16,32,64,128,256,512$
```
                                                                           {A,B,C,D,E,F,G,H,I,J} 1023
                                                                                    / \
                                                             {A,B,C,D,E,F,G,H,I} 511  {J} 512
                                                                     / \
                                                {A,B,C,D,E,F,G,H} 255   {I} 256
                                                        / \
                                     {A,B,C,D,E,F,G} 127   {H} 128
                                            / \
                            {A,B,C,D,E,F} 63   {G} 64
                                 / \
                   {A,B,C,D,E} 31   {F} 32
                       / \
           {A,B,C,D} 15   {E} 16
               / \
      {A,B,C} 7   {D} 8
         / \
  {A,B} 3   {C} 4
   / \
{A} 1 {B} 2
```

For general alphabet of *n* symbols:

Bits to encode a symbol = number of *edges traversed*

Most frequent symbol requires $1$ bit
- First leaf node

Least frequency symbol requires $n-1$ bits
- Deepest leaf node